core/src/main/java/org/apache/calcite/rex/RexPermuteInputsShuttle.java

```diff
@@ -85,9 +101,15 @@ private static ImmutableList<RelDataTypeField> fields(RelNode[] inputs) {
   @Override public RexNode visitInputRef(RexInputRef local) {
     final int index = local.getIndex();
     int target = mapping.getTarget(index);
-    return new RexInputRef(
-        target,
-        local.getType());
+    if (!isStrictTypeMatch) {
```

Comment:
this doesn't look to me like it's a "strict" comparison
basically you are copying types from different fields: source, or destination
Am I right?

Reply:
Yes, maybe I need to change the name `isMatchTargetType`? This affects the creation of subsequent Filters in the rule.